### ISSUE TITLE
Fix to setup.py for building on Windows

### ISSUE THREAD

**Issue by enodwell** [Unidata/netcdf4-python#460]
Hi,

I don't know if others have had success building on Windows, but I had to make a couple of minor changes to get it to work:

1. On windows the library name of the zlib library should be specified as "zlib" instead of "z".

2.  runtime_library_dirs should be set to [] .

Note that I compiled against the binary version of netCDF downloaded from http://www.unidata.ucar.edu/software/netcdf/docs/winbin.html . I also grabbed the binary installer for HDF5 1.8.14, but just for the headers that are missing from the netCDF binary install.

Here is a diff for setup.py :+1: 

--- setup.py.original   Fri Sep 18 12:26:27 2015
+++ setup.py    Fri Sep 18 11:47:20 2015
@@ -300,7 +300,10 @@
     if netCDF4_libdir is None and netCDF4_dir is not None:
         netCDF4_libdir = os.path.join(netCDF4_dir, 'lib')

-    libs = ['netcdf','hdf5_hl','hdf5','z']
+    if sys.platform=='win32':
+        libs = ['netcdf','hdf5_hl','hdf5','zlib']
+    else:
+        libs = ['netcdf','hdf5_hl','hdf5','z']
     if netCDF4_libdir is not None: lib_dirs=[netCDF4_libdir]
     if HDF5_libdir is not None: lib_dirs.append(HDF5_libdir)
     if netCDF4_incdir is not None: inc_dirs=[netCDF4_incdir]
@@ -344,6 +347,11 @@
         lib_dirs.append(curl_libdir)
         inc_dirs.append(curl_incdir)

+if sys.platform=='win32':
+    runtime_lib_dirs = []
+else:
+    runtime_lib_dirs = lib_dirs
+
 # Do not require numpy for just querying the package
 # Taken from the h5py setup file.
 if any('--' + opt in sys.argv for opt in Distribution.display_option_names +
@@ -371,7 +379,7 @@
                             libraries=libs,
                             library_dirs=lib_dirs,
                             include_dirs=inc_dirs,
-                            runtime_library_dirs=lib_dirs),
+                            runtime_library_dirs=runtime_lib_dirs),
                   Extension('netcdftime._datetime', ['netcdftime/_datetime.pyx'])]
     # remove netCDF4.c file if it exists, so cython will recompile netCDF4.pyx.
     # run for build *and* install (issue #263). Otherwise 'pip install' will
@@ -407,7 +415,7 @@
                             libraries=libs,
                             library_dirs=lib_dirs,
                             include_dirs=inc_dirs,
-                            runtime_library_dirs=lib_dirs),
+                            runtime_library_dirs=runtime_lib_dirs),
                   Extension('netcdftime._datetime', ['netcdftime/_datetime.c'])]
     ext_modules = extensions

**Comment by enodwell** [comment]
Sorry, didn't account for Markdown on the diff output. Hope fully this will be more clear:

```
--- setup.py.original   Fri Sep 18 12:26:27 2015
+++ setup.py    Fri Sep 18 11:47:20 2015
@@ -300,7 +300,10 @@
     if netCDF4_libdir is None and netCDF4_dir is not None:
         netCDF4_libdir = os.path.join(netCDF4_dir, 'lib')

-    libs = ['netcdf','hdf5_hl','hdf5','z']
+    if sys.platform=='win32':
+        libs = ['netcdf','hdf5_hl','hdf5','zlib']
+    else:
+        libs = ['netcdf','hdf5_hl','hdf5','z']
     if netCDF4_libdir is not None: lib_dirs=[netCDF4_libdir]
     if HDF5_libdir is not None: lib_dirs.append(HDF5_libdir)
     if netCDF4_incdir is not None: inc_dirs=[netCDF4_incdir]
@@ -344,6 +347,11 @@
         lib_dirs.append(curl_libdir)
         inc_dirs.append(curl_incdir)

+if sys.platform=='win32':
+    runtime_lib_dirs = []
+else:
+    runtime_lib_dirs = lib_dirs
+
 # Do not require numpy for just querying the package
 # Taken from the h5py setup file.
 if any('--' + opt in sys.argv for opt in Distribution.display_option_names +
@@ -371,7 +379,7 @@
                             libraries=libs,
                             library_dirs=lib_dirs,
                             include_dirs=inc_dirs,
-                            runtime_library_dirs=lib_dirs),
+                            runtime_library_dirs=runtime_lib_dirs),
                   Extension('netcdftime._datetime', ['netcdftime/_datetime.pyx'])]
     # remove netCDF4.c file if it exists, so cython will recompile netCDF4.pyx.
     # run for build *and* install (issue #263). Otherwise 'pip install' will
@@ -407,7 +415,7 @@
                             libraries=libs,
                             library_dirs=lib_dirs,
                             include_dirs=inc_dirs,
-                            runtime_library_dirs=lib_dirs),
+                            runtime_library_dirs=runtime_lib_dirs),
                   Extension('netcdftime._datetime', ['netcdftime/_datetime.c'])]
     ext_modules = extensions

```

**Comment by jswhit** [comment]
@cgohlke, does this look reasonable to you?

**Comment by cgohlke** [comment]
Looks reasonable to me. However the actual libraries needed will depend on static vs dynamic linking and the features used to compile NetCDF and HDF.

**Comment by enodwell** [comment]
That is quite true about the libraries. It is true on any system. Currently the setup.py has a hard-coded list of libraries, and the list of libraries that I gave matches the Windows netcdf binary as downloaded from unidata. Perhaps the setup.py script could be more flexible but I know how these things go... it will be done eventually :)

The more important change is clearing runtime_library_dirs. I admit I find it paradoxical that Windows is the only system I know where there is a distinction between link and runtime libraries, and yet it is precisely on Windows that distrib chokes if you specify anything at all for runtime_library_dirs. Here is the resulting error if you do:

```
warning: I don't know what to do with 'runtime_library_dirs': ['C:\\Users\\Eric\\Install\\netCDF-4.3.3.1\\lib', 'C:\\Users\\Eric\\Install\\HDF_Group\\HDF5\\1.8.14\\lib']
error: don't know how to set runtime library search path for MSVC++
```

Perhaps the philosophy of distutils is: "My job ends at link time. Run time is your problem". It could be that there is a way to set runtime_library_dirs usefully, and someone with more Windows knowledge could comment, but in my experience, it needs to be an empty list on Windows.

Eric

**Comment by jswhit** [comment]
would you mind creating a pull request for this?

**Comment by jswhit** [comment]
the runtime_library_dirs issue might be related to this:

http://bugs.python.org/issue2437

**Comment by enodwell** [comment]
I don't think that is the issue, as it is mentioned as a MinGW or Cygwin issue, and I am running Visual Studio.  Details:

Visual Studio 2013 (for good measure I also tried VS 2015)
Windows 10, 64bit
Python 2.7.10, amd64

In my opinion - and I am not a Windows expert, doing most of my development on Linux and OS-X -  this has to do with the fact that there is no Windows equivalent to an RPATH (i.e. no way to encode in a binary a path to needed libraries). Thus on Windows one typically either installs the required libraries to a system path (fortunately not so common anymore, as it led to "DLL hell"), or else one puts the necessary DLLs in the same directory as the executable (the location of the current executable being always implicitly "on the path" in Windows), or else one modifies the system or user PATH variable. Anyway, and this is purely speculation, my guess is that runtime_library_dirs is intended as a way to set RPATH, and thus fails on Windows. This seems to be what is stated at https://docs.python.org/2/distutils/apiref.html

"runtime_library_dirs is a list of directories that will be embedded into the shared library and used to search for other shared libraries that *it* depends on at run-time. (This may only be relevant on Unix.)"

Sure, I can do a pull request.

**Comment by jswhit** [comment]
I went ahead and created a pull request (pull request #461)

**Comment by jswhit** [comment]
merged pull request #461, closing now...

**Comment by enodwell** [comment]
Great thanks.
